module/executiondatasync/provider/provider.go

```diff
@@ -0,0 +1,234 @@
+	blobService network.BlobService,
+	storage tracker.Storage,
+	opts ...ProviderOption,
+) *Provider {
+	p := &Provider{
+		logger:      logger.With().Str("component", "execution_data_provider").Logger(),
+		metrics:     metrics,
+		maxBlobSize: execution_data.DefaultMaxBlobSize,
+		serializer:  serializer,
+		blobService: blobService,
+		storage:     storage,
+	}
+
+	for _, opt := range opts {
+		opt(p)
+	}
+
+	return p
+}
+
+func (p *Provider) storeBlobs(parent context.Context, blockHeight uint64, blobCh <-chan blobs.Blob) <-chan error {
+	ch := make(chan error, 1)
+	go func() {
+		defer close(ch)
+
+		start := time.Now()
+
+		var blobs []blobs.Blob
+		var cids []cid.Cid
+		var totalSize uint64
+		for blob := range blobCh {
+			blobs = append(blobs, blob)
+			cids = append(cids, blob.Cid())
+			totalSize += uint64(len(blob.RawData()))
+		}
+
+		cidArr := zerolog.Arr()
+		for _, cid := range cids {
+			cidArr = cidArr.Str(cid.String())
+		}
+		p.logger.Debug().Array("cids", cidArr).Uint64("height", blockHeight).Msg("storing blobs")
+
+		err := p.storage.Update(func(trackBlobs tracker.TrackBlobsFn) error {
+			ctx, cancel := context.WithCancel(parent)
+			defer cancel()
+
+			if err := trackBlobs(blockHeight, cids...); err != nil {
+				return fmt.Errorf("failed to track blobs: %w", err)
+			}
+
+			if err := p.blobService.AddBlobs(ctx, blobs); err != nil {
+				return fmt.Errorf("failed to add blobs: %w", err)
+			}
+
+			return nil
+		})
+		duration := time.Since(start)
+
+		if err != nil {
+			ch <- err
+			p.metrics.AddBlobsFailed()
+		} else {
+			p.metrics.AddBlobsSucceeded(duration, totalSize)
+		}
+	}()
+
+	return ch
+}
+
+// Provide adds the block execution data for a newly executed (generally not sealed or finalized) block to the blob store for distribution using Bitswap.
+// It computes and returns the root CID of the execution data blob tree.
+func (p *Provider) Provide(ctx context.Context, blockHeight uint64, executionData *execution_data.BlockExecutionData) (flow.Identifier, error) {
+	rootID, errCh, err := p.provide(ctx, blockHeight, executionData)
+	storeErr, ok := <-errCh
+
+	if err != nil {
+		return flow.ZeroID, err
+	}
+
+	if ok {
+		return flow.ZeroID, storeErr
+	} else {
+		return rootID, nil
+	}
+}
+
+func (p *Provider) provide(ctx context.Context, blockHeight uint64, executionData *execution_data.BlockExecutionData) (flow.Identifier, <-chan error, error) {
+	logger := p.logger.With().Uint64("height", blockHeight).Str("block_id", executionData.BlockID.String()).Logger()
+	logger.Debug().Msg("providing execution data")
+
+	start := time.Now()
+
+	blobCh := make(chan blobs.Blob)
+	defer close(blobCh)
+
+	errCh := p.storeBlobs(ctx, blockHeight, blobCh)
+	g, gCtx := errgroup.WithContext(ctx)
+
+	chunkDataIDs := make([]cid.Cid, len(executionData.ChunkExecutionDatas))
+	for i, chunkExecutionData := range executionData.ChunkExecutionDatas {
+		i := i
+		chunkExecutionData := chunkExecutionData
+
+		g.Go(func() error {
```

Comment:
nit: depending on the possible number of chunks in the executionData you may consider upperbounding concurrency here.  (Or ideally, creating a worker pool for the `Provider` during `New`)